protocols/relay/src/priv_client/handler.rs

```diff
@@ -458,13 +357,139 @@ impl ConnectionHandler for Handler {
             return Poll::Ready(ConnectionHandlerEvent::Close(err));
         }
 
+        // Reservations
+        if let Poll::Ready((_, worker_res)) = self.reserve_futs.poll_unpin(cx) {
+            if worker_res.is_err() {
+                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
+            }
+
+            let event = match worker_res.unwrap() {
+                Ok(outbound_hop::Output::Reservation {
+                    renewal_timeout,
+                    addrs,
+                    limit,
+                    to_listener,
+                }) => ConnectionHandlerEvent::NotifyBehaviour(self.reservation.accepted(
+                    renewal_timeout,
+                    addrs,
+                    to_listener,
+                    self.local_peer_id,
+                    limit,
+                )),
+                Err(err) => match err {
+                    outbound_hop::UpgradeError::ReservationFailed(e) => {
+                        let renewal = self.reservation.failed();
+                        return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                            Event::ReservationReqFailed {
+                                renewal,
+                                error: StreamUpgradeError::Apply(e),
+                            },
+                        ));
+                    }
+                    outbound_hop::UpgradeError::Fatal(e) => {
+                        ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(Either::Right(e)))
+                    }
+                    outbound_hop::UpgradeError::CircuitFailed(_) => {
+                        unreachable!("do not emit `CircuitFailed` for reservation")
+                    }
+                },
+                _ => unreachable!("do not emit 'Output::Circuit' for reservation"),
+            };
+
+            return Poll::Ready(event);
+        }
+
+        // Circuit connections
+        if let Poll::Ready((_, worker_res)) = self.circuit_connection_futs.poll_unpin(cx) {
+            if worker_res.is_err() {
+                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
+            }
+
+            let opt = match worker_res.unwrap() {
```

Comment:
Instead of the `unwrap`, would a `match worker_res` work here?

```suggestion
            let opt = match worker_res {
                Err(the-concrete-type-here) => {
                    return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
                },
                Ok(opt) => opt,
            };
```

Reply:
Sure. 
Initially, I did like you suggested, but thought that this way looks a little shorter.